retworkx-core/src/connectivity/conn_components.rs

```diff
@@ -0,0 +1,222 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use hashbrown::HashSet;
+use std::collections::VecDeque;
+use std::hash::Hash;
+
+use petgraph::visit::{GraphProp, IntoNeighbors, IntoNodeIdentifiers, VisitMap, Visitable};
+
+/// Given an graph, a node in the graph, and a visit_map,
+/// return the set of nodes connected to the given node.
+///
+/// Arguments:
+///
+/// * `graph` - The graph object to run the algorithm on
+/// * `node` - The node index to find the connected nodes for
+/// * `graph.visit_map()` - The visit map for the graph
+///
+/// # Example
+/// ```rust
+/// use std::iter::FromIterator;
+/// use hashbrown::HashSet;
+/// use petgraph::graph::Graph;
+/// use petgraph::graph::node_index as ndx;
+/// use petgraph::visit::Visitable;
+/// use petgraph::Directed;
+/// use retworkx_core::connectivity::bfs_undirected;
+///
+/// let graph = Graph::<(), (), Directed>::from_edges(&[
+///     (0, 1),
+///     (1, 2),
+///     (2, 3),
+///     (3, 0),
+///     (4, 5),
+///     (5, 6),
+///     (6, 7),
+///     (7, 4),
+/// ]);
+/// let node_idx = ndx(3);
+/// let component = bfs_undirected(&graph, node_idx, &mut graph.visit_map());
+/// let expected = HashSet::from_iter([ndx(0), ndx(1), ndx(3), ndx(2)]);
+/// assert_eq!(expected, component);
+/// ```
+pub fn bfs_undirected<G>(graph: G, start: G::NodeId, discovered: &mut G::Map) -> HashSet<G::NodeId>
+where
+    G: GraphProp + IntoNeighbors + Visitable,
+    G::NodeId: Eq + Hash,
+{
+    let mut component = HashSet::new();
+    component.insert(start);
+    let mut stack = VecDeque::new();
+    stack.push_front(start);
+
+    while let Some(node) = stack.pop_front() {
+        for succ in graph.neighbors(node) {
```

Comment:
It's not as simple as dropping `GraphProp<EdgeType = Undirected>` restriction. It will give wrong output for directed graphs since the algorithm that finds the weakly connected components of a directed graph must visit the neighbors of a node in *both* directions (incoming and outgoing).

For example the following fails (ideally our retworkx tests should have caught this):

```python
import retworkx

graph = retworkx.PyDiGraph()
graph.add_nodes_from(range(4))
graph.add_edges_from_no_data([
    (3, 2), (2, 1), (1, 0)
])

comps = retworkx.weakly_connected_components(graph)
# Graph is weakly connected, so it should have just one component
assert len(comps) == 1
```

There are some ways to fix this:

1. We can keep the old `retworkx` version and restrict `GraphProp<EdgeType = Undirected>` in order to accept only undirected graphs in `retworkx-core`, or
2. We can restrict `G: IntoNeighborsDirected` and:
```suggestion
        for succ in graph
            .neighbors_directed(node, Outgoing)
            .chain(graph.neighbors_directed(node, Incoming))
        {
```